File: wallet-kit/src/main/java/bitcoin/walllet/kit/network/message/GetBlocksMessage.java
package bitcoin.walllet.kit.network.message;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.Arrays;

import bitcoin.walllet.kit.common.constant.BitcoinConstants;
import bitcoin.walllet.kit.common.io.BitcoinInput;
import bitcoin.walllet.kit.common.io.BitcoinOutput;
import bitcoin.walllet.kit.common.util.HashUtils;

/**
 * Build P2P message:
 * https://en.bitcoin.it/wiki/Protocol_documentation#Message_structure
 *
 * @author liaoxuefeng
 */
public class GetBlocksMessage extends Message {

    int version; // uint32
    byte[][] hashes; // byte[32]
    byte[] hashStop; // hash of the last desired block header; set to zero to
    // get as many blocks as possible (2000)

    public GetBlocksMessage(byte[] firstHash, byte[] hashStop) {
        super("getblocks");
        this.version = BitcoinConstants.PROTOCOL_VERSION;
        this.hashes = new byte[][]{firstHash};
        this.hashStop = hashStop;
    }

    public GetBlocksMessage(byte[] payload) throws IOException {
        super("getblocks");
        try (BitcoinInput input = new BitcoinInput(new ByteArrayInputStream(payload))) {
            this.version = input.readInt();
            long hashCount = input.readVarInt(); // do not keep hash count
            this.hashes = new byte[(int) hashCount][];
            for (int i = 0; i < hashCount; i++) {
                this.hashes[i] = input.readBytes(32);
            }
            this.hashStop = input.readBytes(32);
        }
    }

    @Override
    protected byte[] getPayload() {
        BitcoinOutput output = new BitcoinOutput();
        output.writeInt(this.version).writeVarInt(this.hashes.length);
        for (int i = 0; i < this.hashes.length; i++) {
            output.write(this.hashes[i]);
        }
        output.write(this.hashStop);
        return output.toByteArray();
    }

    @Override
    public String toString() {
        return "GetBlocksMessage(" + this.hashes.length + ": ["
                + String.join(", ", Arrays.stream(this.hashes).map((hash) -> {
            return HashUtils.toHexStringAsLittleEndian(hash);
        }).limit(10).toArray(String[]::new)) + "], hashStop="
                + HashUtils.toHexStringAsLittleEndian(this.hashStop) + ")";
    }

}

File: wallet-kit/src/main/kotlin/bitcoin/wallet/kit/messages/GetBlocksMessage.kt
package bitcoin.wallet.kit.messages

import bitcoin.walllet.kit.common.constant.BitcoinConstants
import bitcoin.walllet.kit.common.io.BitcoinInput
import bitcoin.walllet.kit.common.io.BitcoinOutput
import bitcoin.walllet.kit.common.util.HashUtils
import java.io.ByteArrayInputStream
import java.io.IOException

/**
 * GetBlocks Message
 *
 *  Size       Field       Description
 *  ====       =====       ===========
 *  4 bytes    Version     Negotiated protocol version
 *  VarInt     Count       Number of locator hash entries
 *  Variable   Entries     Locator hash entries
 *  32 bytes   Stop        Hash of the last desired block or zero to get as many as possible
 */
class GetBlocksMessage : Message {

    private var version: Int = 0                    // uint32
    private lateinit var hashes: Array<ByteArray>   // byte[32]
    private lateinit var hashStop: ByteArray        // hash of the last desired block header; set to zero to get as many blocks as possible (2000)

    constructor(firstHash: ByteArray, hashStop: ByteArray) : super("getblocks") {
        version = BitcoinConstants.PROTOCOL_VERSION
        hashes = arrayOf(firstHash)
        this.hashStop = hashStop
    }

    @Throws(IOException::class)
    constructor(payload: ByteArray) : super("getblocks") {
        BitcoinInput(ByteArrayInputStream(payload)).use { input ->
            version = input.readInt()
            val hashCount = input.readVarInt() // do not keep hash count
            hashes = Array(hashCount.toInt()) { input.readBytes(32) }
            hashStop = input.readBytes(32)
        }
    }

    override fun getPayload(): ByteArray {
        val output = BitcoinOutput()
        output.writeInt(version).writeVarInt(hashes.size.toLong())
        for (i in hashes.indices) {
            output.write(hashes[i])
        }
        output.write(hashStop)
        return output.toByteArray()
    }

    override fun toString(): String {
        val list = hashes
                .take(10)
                .map { hash -> HashUtils.toHexStringAsLittleEndian(hash) }
                .joinToString()

        return ("GetBlocksMessage(" + hashes.size + ": [" + list + "], hashStop=" + HashUtils.toHexStringAsLittleEndian(hashStop) + ")")
    }

}